File: 2017/AoC17-06ab.swift
func redistribute(_ banks: inout [Int]) -> Int {
  var record = Set<[Int]>()

  let len = banks.count
  while record.insert(banks).inserted {
    let (i, e) = banks.enumerated().max(by: { $0.element < $1.element } )!

    banks[i] = 0

    for x in 1...e {
      banks[(i + x) % len] += 1
    }
  }

  return record.count
}

var banks = readLine()!.split(separator: "\t").map( { Int($0)! } )

print("Part 1 answer: \(redistribute(&banks))")
print("Part 2 answer: \(redistribute(&banks))")

File: 2017/AoC17-06ab.rs
use std::fs;
use std::collections::HashSet;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  let mut banks: Vec<u32> = input.split_whitespace().map(|s| s.parse().unwrap()).collect();

  println!("Part 1 answer: {}", redistribute(&mut banks));
  println!("Part 2 answer: {}", redistribute(&mut banks));
}

fn redistribute(banks: &mut [u32]) -> usize {
  let mut record: HashSet<Vec<u32>> = HashSet::new();

  let len = banks.len();
  while record.insert(banks.to_owned()) {
    let (i, e) = banks.to_owned().into_iter().enumerate().rev().max_by_key(|&(_, n)| n).unwrap();

    banks[i] = 0;
    for n in 1..=e as usize {
      banks[(i + n) % len] += 1;
    }
  }

  record.len()
}